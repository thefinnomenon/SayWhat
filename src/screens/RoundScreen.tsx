import React, { useState } from 'react';
import { Theme } from '../../types';
import styled from 'styled-components/native';
import { useNavigation } from 'react-navigation-hooks';
import { useTranslation } from 'react-i18next';
//@ts-ignore
import ConfettiCannon from 'react-native-confetti-cannon';
import Button from '../components/Button';
import Score from '../components/Score';

const SCORE_GOAL = 5;

const initialState = Object.freeze({
  winner: '',
  round: 1,
  teamA: 0,
  teamB: 0,
  scoreWasUpdated: true,
});

export const RoundScreen = () => {
  const [winner, setWinner] = useState(initialState.winner);
  const [round, setRound] = useState(initialState.round);
  const [teamA, setTeamA] = useState(initialState.teamA);
  const [teamB, setTeamB] = useState(initialState.teamB);
  const [scoreWasUpdated, setScoreWasUpdated] = useState(
    initialState.scoreWasUpdated,
  );

  const { navigate } = useNavigation();
  const { t } = useTranslation();

  const updateScore = (team: string) => {
    switch (team) {
      case 'A':
        setTeamA(teamA + 1);
        if (teamA + 1 === SCORE_GOAL) {
          setWinner('A');
        }
        break;
      case 'B':
        setTeamB(teamB + 1);
        if (teamB + 1 === SCORE_GOAL) {
          setWinner('B');
        }
    }

    setScoreWasUpdated(true);
  };

  return (
    <>
      <Container>
        {winner !== '' && (
          <ConfettiCannon count={200} origin={{ x: 0, y: 0 }} />
        )}
        <TitleContainer>
          <Title>{`Round ${round}`}</Title>
        </TitleContainer>
        <ScoreContainer>
          <Score
            teamName="A"
            score={teamA}
            disabled={scoreWasUpdated}
            onPress={() => updateScore('A')}
          />
          <Spacer flex={1} />
          <Score
            teamName="B"
            score={teamB}
            disabled={scoreWasUpdated}
            onPress={() => updateScore('B')}
          />
        </ScoreContainer>
        <Spacer flex={2} />
      </Container>
      <BottomButton
        disabled={!scoreWasUpdated}
        title="START ROUND"
        onPress={() => navigate('Game')}
      />
    </>
  );
};

const Container = styled.View<Theme>`
  flex: 1;
  justify-content: space-between;
  align-items: center;
  background: ${props => props.theme.colors.background};
`;

const TitleContainer = styled.View`
  flex: 1;
  justify-content: center;
  align-items: center;
`;

const ScoreContainer = styled.View<Theme>`
  flex: 1;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  width: 80%;
`;

const Title = styled.Text<Theme>`
  color: ${props => props.theme.colors.text};
  font-size: 48;
  font-weight: bold;
  margin: 8px;
`;

const BottomButton = styled(Button)<Theme>`
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;
  height: 80px;
  background: ${props => props.theme.colors.primary};
`;

type SpacerProps = {
  flex: number;
};
const Spacer = styled.View<SpacerProps>`
  flex: ${props => props.flex};
`;

RoundScreen.navigationOptions = {
  title: '',
};

export default RoundScreen;
