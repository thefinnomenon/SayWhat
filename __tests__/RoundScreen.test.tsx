import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import ThemeManager from '../contexts/ManageThemeContext';
import { I18nextProvider } from 'react-i18next';
import i18 from '../i18n';
import RoundScreen from '../src/screens/RoundScreen';

const renderComponent = () =>
  renderer
    .create(
      <ThemeManager>
        <I18nextProvider i18n={i18}>
          <RoundScreen />
        </I18nextProvider>
      </ThemeManager>,
    )
    .toJSON();

it('renders correctly', () => {
  const tree = renderComponent();
  expect(tree).toMatchSnapshot();
});
