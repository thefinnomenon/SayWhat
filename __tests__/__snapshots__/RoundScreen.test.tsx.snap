// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <Modal
      animationType="slide"
      hardwareAccelerated={false}
      onRequestClose={[Function]}
      transparent={true}
      visible={false}
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#e8e8e8",
              "borderRadius": 20,
              "height": "25%",
              "justifyContent": "center",
              "marginBottom": "auto",
              "marginLeft": "auto",
              "marginRight": "auto",
              "marginTop": "auto",
              "paddingBottom": 8,
              "paddingLeft": 8,
              "paddingRight": 8,
              "paddingTop": 8,
              "width": "80%",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#202020",
                "fontSize": 80,
              },
            ]
          }
        >
          Team  wins!
        </Text>
      </View>
    </Modal>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#202020",
              "fontSize": 96,
              "fontWeight": "bold",
              "marginBottom": 16,
              "marginLeft": 16,
              "marginRight": 16,
              "marginTop": 16,
            },
          ]
        }
      >
        Round 1
      </Text>
    </View>
    <View
      flex={1}
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexBasis": 0,
            "flexDirection": "row",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "space-between",
            "width": "80%",
          },
        ]
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#50b7ff",
            "borderRadius": 20,
            "flexBasis": 0,
            "flexGrow": 4,
            "flexShrink": 1,
            "height": 360,
            "justifyContent": "center",
            "opacity": 1,
            "paddingBottom": 8,
            "paddingLeft": 8,
            "paddingRight": 8,
            "paddingTop": 8,
            "width": "80%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#fff",
                "fontSize": 48,
                "fontWeight": "bold",
                "marginBottom": 16,
                "marginLeft": 16,
                "marginRight": 16,
                "marginTop": 16,
                "textAlign": "center",
              },
            ]
          }
        >
          A
        </Text>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#fff",
                "borderRadius": 20,
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
                "justifyContent": "center",
                "marginBottom": 20,
                "marginLeft": 20,
                "marginRight": 20,
                "marginTop": 20,
                "width": "85%",
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#202020",
                  "fontSize": 96,
                  "textAlign": "center",
                },
              ]
            }
          >
            0
          </Text>
        </View>
      </View>
      <View
        flex={1}
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      />
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#50b7ff",
            "borderRadius": 20,
            "flexBasis": 0,
            "flexGrow": 4,
            "flexShrink": 1,
            "height": 360,
            "justifyContent": "center",
            "opacity": 1,
            "paddingBottom": 8,
            "paddingLeft": 8,
            "paddingRight": 8,
            "paddingTop": 8,
            "width": "80%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#fff",
                "fontSize": 48,
                "fontWeight": "bold",
                "marginBottom": 16,
                "marginLeft": 16,
                "marginRight": 16,
                "marginTop": 16,
                "textAlign": "center",
              },
            ]
          }
        >
          B
        </Text>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#fff",
                "borderRadius": 20,
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
                "justifyContent": "center",
                "marginBottom": 20,
                "marginLeft": 20,
                "marginRight": 20,
                "marginTop": 20,
                "width": "85%",
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#202020",
                  "fontSize": 96,
                  "textAlign": "center",
                },
              ]
            }
          >
            0
          </Text>
        </View>
      </View>
    </View>
    <View
      flex={3}
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexGrow": 3,
            "flexShrink": 1,
          },
        ]
      }
    />
  </View>,
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignContent": "center",
        "backgroundColor": "#50b7ff",
        "borderRadius": 0,
        "bottom": 0,
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "height": 150,
        "justifyContent": "center",
        "left": 0,
        "opacity": 1,
        "position": "absolute",
        "right": 0,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#fff",
            "fontSize": 40,
            "fontWeight": "bold",
            "textAlign": "center",
          },
        ]
      }
    >
      START GAME
    </Text>
  </View>,
]
`;
